accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,375 @@
+# Dart Static Extension Methods Design
+lrn@google.com
+
+This is a design document for *static extension methods* for Dart. This document describes the most basic variant of the feature, then lists a few possible variants or extensions.
+
+The design of this feature is kept deliberately simple.
+
+## What Are Static Extension Methods
+
+Dart classes have virtual methods. An invocation like `thing.doStuff()` will invoke the virtual `doStuff` method on the object denoted by `thing`. The only way to add methods to a class is to modify the class. If you are not the author of the class, you have to use static helper functions instead of methods, so `doMyStuff(thing)` instead of `thing.doMyStuff()`. That's acceptable for a single function, but in a larger chain of operations, it becomes overly cumbersome. Example:
+
+```dart
+doMyOtherStuff(doMyStuff(something.doStuff()).doOtherStuff())
+```
+
+That code is so much less readable than
+
+```dart
+something.doStuff().doMyStuff().doOtherStuff().doMyOtherStuff()
+```
+
+So, primarily for readability reasons, static extension methods will allow you to add "extension methods", which are really just static functions, to existing types, and allow you to call those methods as if they were actual methods, using `.`.
+
+The extension methods are *static*, which means that we use the static type of an expression to figure out which method to call, and that also means that static extension methods are not virtual methods.
+
+The methods we allow you to add this way are normal methods, operators, getter and setters.
+
+## Declaring Static Extension Methods
+
+### Syntax
+
+A static extension of a type is declared using syntax like:
+
+```dart
+extension MyFancyList<T> on List<T> {
+  int get doubleLength => this.length * 2;
+  List<T> operator-() => this.reversed.toList();
+  List<List<T>> split(int at) => 
+      <List<T>>[this.sublist(0, at), this.sublist(at)];
+  List<T> mapToList<R>(R Function(T) convert) => this.map(convert).toList();
+}
+```
+
+More precisely, an extension declaration is a top-level declaration with a grammar similar to:
+
+```ebnf
+<extension> ::= 
+  `extension' <identifier> <typeParameters>? `on' <type> `?'? `{'
+     memberDeclaration*
+  `}'
+```
+
+Such a declaration introduces its *name* (the identifier) into the surrounding scope. The name does not denote a type, but it can be used to denote the extension itself in various places. The name can be hidden or shown in `import` or `export` declarations.
+
+The *type* can be any valid Dart type, but not a single type variable. It can refer to the type parameters of the extension. It can be followed by `?` which means that it allows `null` values.
```

Comment:
Erik also couldn't find any reason not to, and I don't remember why I was worried about it.
It's otherwise equivalent to `on Object` (or on the bound of the type variable), but you get the static receiver type as a type. That should be simpler than having to deconstruct the receiver type.

Reply:
The other option I've kicked around is always providing the self type.  e.g., we just say that there's always an implicit `Self` type variable in scope that is the matched type.  Then, as you say, you could just put it on `Object`.  I think for this use case it's simpler to just allow `T` as a pattern.  What I'm not sure of is whether 

```
extension<T> on TypePattern {}
```

applies to exactly the same set of things as:

```
extension<T extends TypePattern> on T {}
```

It seems like it ought to?  And if so, you can always get the self type.  But if not, then you can't both get the self type and have a specific pattern.

Reply:
It should work on the same things, yes.

If `foo(TypePattern x) {}; foo(expression);` is allowed, then so is `foo<T extends TypePattern>(T x) {}; foo(expression);` for exactly the same expressions. We use exactly the same kind of type inference to match the `on` type as we do for such a function invocation.

As for getting both, I had hoped that the following would work:
```dart
extension Foo<E, S extends List<E>> on S {
   S get self => this;
   List<E> clone() => List<E>.of(this);
}
```
but I seem to get `dynamic` for `E` when I try this with a function. :frowning_face:

Reply:
Here's a version that uses a wrapper object rather than a function (I suspect that this will be an even more faithful model for how type inference is performed, given the rest of this document):

```dart
class Foo<E, S extends List<E>> {
  S _self;
  Foo(this._self);
  S get self => _self;
  List<E> clone() => List<E>.of(_self);
}

main() {
  List<int> xs = [1];
  // Emulate extension method invocation `xs.clone()`.
  // At first, use the direct desugaring as wrapper:
  var ys = Foo(xs).clone();
  print(ys.runtimeType); // 'List<dynamic>'.
}
```

In this situation we can't expect type inference to apply any constraints to the first type argument of the `Foo` instance creation (`E`), so it becomes `dynamic`. It would surely be a massively breaking change (and not particularly useful) to "optimize" `E` to `int` just because the second type argument has been inferred to be `List<int>` (assuming that we would do so for type inference everywhere, in order to make sure that we keep the set of special exceptions for extension methods as small as possible).

I believe that the design will satisfy Leaf's rule about applicability ('applies to exactly the same'), but also that we shouldn't expect inference to work the same for `extension<T> on TypePattern {}` and for `extension<T extends TypePattern> on T {}`. So I'd recommend that we just don't aim for having this property, such that we can use the normal rules for type inference with extensions.